x-pack/filebeat/input/s3/input.go

```diff
@@ -0,0 +1,432 @@
+func (c *channelContext) Done() <-chan struct{}       { return c.done }
+func (c *channelContext) Err() error {
+	select {
+	case <-c.done:
+		return context.Canceled
+	default:
+		return nil
+	}
+}
+func (c *channelContext) Value(key interface{}) interface{} { return nil }
+
+// NewInput creates a new s3 input
+func NewInput(cfg *common.Config, outletFactory channel.Connector, context input.Context) (input.Input, error) {
+	cfgwarn.Beta("s3 input type is used")
+	logger := logp.NewLogger(inputName)
+
+	config := defaultConfig()
+	if err := cfg.Unpack(&config); err != nil {
+		return nil, errors.Wrap(err, "failed unpacking config")
+	}
+
+	outlet, err := outletFactory(cfg, context.DynamicFields)
+	if err != nil {
+		return nil, err
+	}
+
+	awsConfig, err := awscommon.GetAWSCredentials(config.AwsConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "getAWSCredentials failed")
+	}
+
+	closeChannel := make(chan struct{})
+	p := &Input{
+		outlet:    outlet,
+		config:    config,
+		awsConfig: awsConfig,
+		logger:    logger,
+		close:     closeChannel,
+		context:   &channelContext{closeChannel},
+	}
+	return p, nil
+}
+
+// Run runs the input
+func (p *Input) Run() {
+	p.workerOnce.Do(func() {
+		visibilityTimeout := int64(p.config.VisibilityTimeout.Seconds())
+		regionName, err := getRegionFromQueueURL(p.config.QueueURL)
+		if err != nil {
+			p.logger.Errorf("failed to get region name from queueURL: %v", p.config.QueueURL)
+		}
+
+		awsConfig := p.awsConfig.Copy()
+		awsConfig.Region = regionName
+		svcSQS := sqs.New(awsConfig)
+		svcS3 := s3.New(awsConfig)
+
+		go p.run(svcSQS, svcS3, visibilityTimeout)
+	})
+}
+
+func (p *Input) run(svcSQS *sqs.Client, svcS3 *s3.Client, visibilityTimeout int64) {
+	defer p.logger.Infof("S3 input worker for '%v' has stopped.", p.config.QueueURL)
+	p.logger.Infof("s3 input worker has started. with queueURL: %v", p.config.QueueURL)
+	for p.context.Err() == nil {
+		// receive messages from sqs
+		req := svcSQS.ReceiveMessageRequest(
+			&sqs.ReceiveMessageInput{
+				QueueUrl:              &p.config.QueueURL,
+				MessageAttributeNames: []string{"All"},
+				MaxNumberOfMessages:   &maxNumberOfMessage,
+				VisibilityTimeout:     &visibilityTimeout,
+				WaitTimeSeconds:       &waitTimeSecond,
+			})
+
+		output, err := req.Send(p.context)
+		if err != nil {
+			p.logger.Error("failed to receive message from SQS:", err)
+			time.Sleep(time.Duration(waitTimeSecond) * time.Second)
+			continue
+		}
+
+		if output == nil || len(output.Messages) == 0 {
+			p.logger.Debug("no message received from SQS:", p.config.QueueURL)
+			continue
+		}
+
+		// process messages received from sqs, get logs from s3 and create events
+		p.processor(p.config.QueueURL, output.Messages, visibilityTimeout, svcS3, svcSQS)
+	}
+}
+
+// Stop stops the s3 input
+func (p *Input) Stop() {
+	defer p.outlet.Close()
+	close(p.close)
+	p.logger.Info("Stopping s3 input")
+}
+
+// Wait stops the s3 input.
+func (p *Input) Wait() {
+	p.Stop()
+}
```

Comment:
QQ: does Wait block?

Reply:
@kaiyan-sheng 	Can you clearify here? The `Stop` method must signal shutdown, but not block. The `Wait` method must signal shutdown (if not already done otherwise) block until shutdown has been completed. I know it's some idiocracy we should improve/remove, but the way the runner currently wraps an input doesn't make this easy for us.

Reply:
Hmm sorry I'm still trying to understand the difference. Does this mean `Stop` method stops the input. But `Wait` method not only needs to stop the input, it also needs to be blocked to make sure all the channels are closed for example?

Reply:
> Does this mean Stop method stops the input. But Wait method not only needs to stop the input, it also needs to be blocked to make sure all the channels are closed for example?

Correct. Other inputs achieve this by introducing a WaitGroup for example.

Reply:
Great, thanks! I just added a `workerWg` to do so.